scheduler/replica_scheduler.go

```diff
@@ -36,22 +36,23 @@ func (rcs *ReplicaScheduler) ScheduleReplica(replica *longhorn.Replica, replicas
 	if len(nodeInfo) == 0 {
 		return nil, fmt.Errorf("There's no available node for replica %v", replica)
 	}
-	// TODO Need to add capacity.
-	// Just make sure replica of the same volume be scheduled to different nodes for now.
-	preferredNodes, err := rcs.preferredNodes(nodeInfo, replicas)
+
+	// find proper node and disk
+	preferredNodes, err := rcs.preferredNodes(nodeInfo, replicas, replica)
```

Comment:
It's no longer preferredNodes. It's all the possible node candidates. And it contains disk information as well.

`diskCandidates, err := rcs.chooseDiskCandidates(nodeInfo, replicas, replica)`

The function can just return a `map[id]*disk`. You don't need to worry about which node to choose since it's either preferred or not, all included in the result.